editor/src/messages/portfolio/document/navigation/navigation_message_handler.rs

```diff
@@ -19,35 +19,43 @@ use serde::{Deserialize, Serialize};
 pub struct NavigationMessageHandler {
 	pub pan: DVec2,
 	panning: bool,
+	pre_commit_pan: DVec2,
 	snap_tilt: bool,
 	snap_tilt_released: bool,
 
 	pub tilt: f64,
+	pre_commit_tilt: f64,
 	tilting: bool,
 
 	pub zoom: f64,
+	pre_commit_zoom: f64,
 	zooming: bool,
 	snap_zoom: bool,
 
 	mouse_position: ViewportPosition,
+	dispatched_from_menu: bool,
 }
 
 impl Default for NavigationMessageHandler {
 	fn default() -> Self {
 		Self {
 			pan: DVec2::ZERO,
 			panning: false,
+			pre_commit_pan: DVec2::ZERO,
 			snap_tilt: false,
 			snap_tilt_released: false,
 
 			tilt: 0.,
+			pre_commit_tilt: 0.,
 			tilting: false,
 
 			zoom: 1.,
+			pre_commit_zoom: 0.,
 			zooming: false,
 			snap_zoom: false,
 
 			mouse_position: ViewportPosition::default(),
+			dispatched_from_menu: false,
```

Comment:
I'm not sure having this as state for the message handler - it seems like it should just be for each message?

Reply:
unfortunately I have to use this in bool in the actions function to enable this to take priority over the currently active tool when starting a translation from outside the viewport. Is there some other way I could thread the state to the action function?

Reply:
I see.